### PR TITLE
[5244][IMP] stock_move_actual_date: improve inventory adjustment accounting…

### PR DESCRIPTION
**Pull request by AungKoKoLin1997**
[5244](https://www.quartile.co/web#id=5244&cids=3&model=project.task&view_type=form)

### CODE REVIEW

File: stock_move_actual_date/models/stock_move.py [hunk 1]
```diff
@@ -51,10 +58,9 @@ def _prepare_account_move_vals(
             svl_id,
             cost,
         )
-        # i.e. Inventory adjustments with actual date
-        if self._context.get("force_period_date"):
-            self.write({"actual_date": self._context["force_period_date"]})
-            return am_vals
+        # # i.e. Inventory adjustments with actual date
+        # if self._context.get("force_period_date"):
+        #     self.write({"actual_date": self._context["force_period_date"]})
```
Comment (yostashiro) [inline stock_move_actual_date/models/stock_move.py]:
```suggestion
```

**Review by yostashiro — COMMENTED** [review]
Please update the commit message with the reason of the change.

**Comment by yostashiro** [comment]
@AungKoKoLin1997 Please check the test error. Did you locally test the functionality?

File: stock_move_actual_date/models/stock_move.py [hunk 2]
```diff
@@ -32,6 +32,13 @@ def _compute_actual_date(self):
                 self.with_context(tz=tz), rec.date
             )
 
+    def _action_done(self, cancel_backorder=False):
+        moves = super()._action_done(cancel_backorder)
+        # i.e. Inventory adjustments with actual date
+        if self._context.get("force_period_date"):
+            self.write({"actual_date": self._context["force_period_date"]})
```
Comment (yostashiro) [inline stock_move_actual_date/models/stock_move.py]: This needs to go before `super()`, or the value is not passed to the account move. Apply filtering on `self` as appropriate.

**Comment by AungKoKoLin1997** [comment]
> @AungKoKoLin1997 Please check the test error. Did you locally test the functionality?

Let me check test error. I tested functionality in my local env.

File: stock_move_actual_date/models/stock_move.py (hunk 2 above)
Comment (AungKoKoLin1997) [reply stock_move_actual_date/models/stock_move.py]: We can't go before `super()` because compute method will override the `actual_date` when it updates the date with current datetime during `action_done` process.

Comment (yostashiro) [reply stock_move_actual_date/models/stock_move.py]: @AungKoKoLin1997 Pass the `actual_date` context, then? Please suggest a good optimal solution.

Comment (AungKoKoLin1997) [reply stock_move_actual_date/models/stock_move.py]: I thought about using context in compute method but rousseldenis mentioned he doesn't like computed stored fields based on context values. So, community may not accept this approach. Let me find the optimal solution.

File: stock_move_actual_date/models/stock_move.py [hunk 3]
```diff
@@ -51,12 +58,8 @@ def _prepare_account_move_vals(
...
             cost,
         )
-        # i.e. Inventory adjustments with actual date
-        if self._context.get("force_period_date"):
-            self.write({"actual_date": self._context["force_period_date"]})
-            return am_vals
-        if self.actual_date:
-            am_vals.update({"date": self.actual_date})
+        actual_date = self._context.get("force_period_date") or self.actual_date
+        am_vals.update({"date": actual_date})
```
Comment (yostashiro) [inline stock_move_actual_date/models/stock_move.py]: To be safe.
```suggestion
        actual_date = self.env.context.get("force_period_date") or self.actual_date
        if actual_date:
            am_vals.update({"date": actual_date})
```

File: stock_move_actual_date/models/stock_move.py [hunk 4]
```diff
@@ -32,6 +32,13 @@ def _compute_actual_date(self):
                 self.with_context(tz=tz), rec.date
             )
 
+    def _action_done(self, cancel_backorder=False):
+        moves = super()._action_done(cancel_backorder)
+        # i.e. Inventory adjustments with actual date
+        if self._context.get("force_period_date"):
```
Comment (yostashiro) [inline stock_move_actual_date/models/stock_move.py]:
```suggestion
        if self.env.context.get("force_period_date"):
```

**Review by yostashiro — COMMENTED** [review]
@AungKoKoLin1997 Can you please add a test for inventory adjustment of a zero-valued product?

File: stock_move_actual_date/models/stock_move.py [hunk 5]
```diff
@@ -32,6 +32,13 @@ def _compute_actual_date(self):
                 self.with_context(tz=tz), rec.date
             )
 
+    def _action_done(self, cancel_backorder=False):
+        moves = super()._action_done(cancel_backorder)
+        # i.e. Inventory adjustments with actual date
+        if self.env.context.get("force_period_date"):
+            self.write({"actual_date": self._context["force_period_date"]})
```
Comment (yostashiro) [inline stock_move_actual_date/models/stock_move.py]:
```suggestion
            self.write({"actual_date": self.env.context["force_period_date"]})
```

**Review by yostashiro — APPROVED** [review]

File: stock_move_actual_date/models/stock_move.py [hunk 6]
```diff
@@ -32,6 +32,13 @@ def _compute_actual_date(self):
                 self.with_context(tz=tz), rec.date
             )
 
+    def _action_done(self, cancel_backorder=False):
+        moves = super()._action_done(cancel_backorder)
+        # i.e. Inventory adjustments with actual date
+        if self.env.context.get("force_period_date"):
+            self.write({"actual_date": self.env.context.get("force_period_date")})
```
Comment (yostashiro) [inline stock_move_actual_date/models/stock_move.py]:
```suggestion
            self.write({"actual_date": self.env.context["force_period_date"]})
```

**Review by yostashiro — COMMENTED** [review]
Sorry, missed one point.

**Review by kanda999 — APPROVED** [review]
Functional test and code review.

**Comment by yostashiro** [comment]
@AungKoKoLin1997 Can you please update the OCA PR and create PRs to update the module in xxx-oca repos?

**Review by nobuQuartile — APPROVED** [review]
Functional test and code review.

File: stock_move_actual_date/security/stock_move_actual_date_security.xml [hunk 7]
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.0" encoding="UTF-8" ?>
+<odoo>
+    <record id="group_actual_date_editable" model="res.groups">
+        <field name="name">Editable Actual Date</field>
```
Comment (yostashiro) [inline stock_move_actual_date/security/stock_move_actual_date_security.xml]:
```suggestion
        <field name="name">Modify Actual Date</field>
```
This sounds better to me as a group name. "Editable Actual Date" sounds a bit misleading.

File: stock_move_actual_date/security/stock_move_actual_date_security.xml [hunk 8]
```diff
@@ -0,0 +1,8 @@
...
+        <field name="category_id" ref="base.module_category_hidden" />
+        <field name="implied_ids" eval="[(4, ref('stock.group_stock_user'))]" />
```
Comment (yostashiro) [inline stock_move_actual_date/security/stock_move_actual_date_security.xml]:
```suggestion
```
I think we can remove this to be a bit more flexible. Or does this group make no sense without this?

Comment (AungKoKoLin1997) [reply stock_move_actual_date/security/stock_move_actual_date_security.xml]: I think it is better to remove because we can't also add another group in this module (eg, `group_mrp_user`).

**Review by yostashiro — APPROVED** [review]
@AungKoKoLin1997 @nobuQuartile Can you please confirm that the existing stock.move `actual_date` values survive for the cases where actual date is set in the picking just in case?

**Comment by AungKoKoLin1997** [comment]
> Can you please confirm that the existing stock.move `actual_date` values survive for the cases where actual date is set in the picking just in case?

The existing stock.move `actual_date` values are updated when actual date is set in the picking.

**Comment by yostashiro** [comment]
@AungKoKoLin1997 I'm talking about the case where actual_date of the picking is already set when we upgrade the module.

**Comment by AungKoKoLin1997** [comment]
> > Can you please confirm that the existing stock.move `actual_date` values survive for the cases where actual date is set in the picking just in case?
> 
> The existing stock.move `actual_date` values are updated when actual date is set in the picking.

I think I misunderstood your question. Except from related moves of picking that is changed `actual_date`, the others are remain the same.

**Comment by AungKoKoLin1997** [comment]
> Can you please confirm that the existing stock.move `actual_date` values survive for the cases where actual date is set in the picking just in case?

 I tested with below scenario and it is working.
 
- Set actual_date with previous code for a picking.
- Update module with latest code and upgrade the module
- The `actual_date` of stock moves remain the same but `acutal_date_source` is empty(It is not the case I think)
- After updated `actual_date` in picking, the actual_date and `actual_date_source` of stock moves are updated.

**Comment by yostashiro** [comment]
@AungKoKoLin1997 How about this case?

- Create a picking with actual date set
- Upgrade the stock_move_actual_date module
- Validate the picking

**Comment by AungKoKoLin1997** [comment]
> How about this case?

@yostashiro It doesn't remain the same and update with date of stock move.
I will figure out how to solve this.

**Comment by AungKoKoLin1997** [comment]
> I will figure out how to solve this.

@yostashiro How about we just add the migration script for assigning `actual_date_source` from actual_date? Hooks will not work when we upgrade the module I think.

**Review by yostashiro — APPROVED** [review]

**Comment by yostashiro** [comment]
@AungKoKoLin1997 Please update this PR with the latest OCA code (or close this and create a new one). High priority.

**Review by yostashiro — APPROVED** [review]
